src/audio/aria/aria.c

```diff
@@ -0,0 +1,348 @@
+	}
+
+	ret = aria_algo_init(dev, buf, buf_in, buf_out, att, chc, sgc);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static struct comp_dev *aria_new(const struct comp_driver *drv,
+				 struct comp_ipc_config *config, void *spec)
+{
+	struct comp_dev *dev;
+	int ret;
+
+	comp_cl_info(&comp_aria, "aria_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	ret = init_aria(dev, config, spec);
+	if (ret < 0) {
+		comp_free(dev);
+		return NULL;
+	}
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static void aria_free(struct comp_dev *dev)
+{
+	struct aria_data *cd = comp_get_drvdata(dev);
+
+	rfree(cd->data_);
+	rfree(cd->buf_in);
+	rfree(cd->buf_out);
+	rfree(cd);
+	rfree(dev);
+}
+
+static int aria_prepare(struct comp_dev *dev)
+{
+	int ret;
+
+	comp_info(dev, "aria_prepare()");
+
+	if (dev->state == COMP_STATE_ACTIVE) {
+		comp_info(dev, "aria_prepare(): Component is in active state.");
+		return 0;
+	}
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	return 0;
+}
+
+static int aria_reset(struct comp_dev *dev)
+{
+	struct aria_data *cd = comp_get_drvdata(dev);
+	int idx;
+	int ret = 0;
+
+	comp_info(dev, "aria_reset()");
+
+	if (dev->state == COMP_STATE_ACTIVE) {
+		comp_info(dev, "aria module is in active state. Ignore resetting");
+		return 0;
+	}
+
+	for (idx = 0; idx < ARIA_MAX_GAIN_STATES; ++idx)
+		cd->gains_[idx] = (1ULL << (32 - cd->att_ - 1)) - 1;
+
+	memset(cd->data_, 0, sizeof(int32_t) * cd->buff_size_);
+	memset(cd->buf_in, 0, sizeof(int32_t) * cd->buff_size_);
+	memset(cd->buf_out, 0, sizeof(int32_t) * cd->buff_size_);
+	cd->gain_state_ = 0;
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	return ret;
+}
+
+static int aria_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "aria_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+static int aria_copy(struct comp_dev *dev)
+{
+	struct comp_copy_limits c;
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+	uint32_t source_bytes;
+	uint32_t sink_bytes;
+	uint32_t buff_frag = 0;
+	int32_t *destp;
+	int32_t i, channel;
+	struct aria_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "aria_copy()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 sink_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
+
+	comp_get_copy_limits_with_lock(source, sink, &c);
+	source_bytes = c.frames * c.source_frame_bytes;
+	sink_bytes = c.frames * c.sink_frame_bytes;
+
+	if (source_bytes == 0)
+		return 0;
+
+	buffer_invalidate(source, source_bytes);
+
+	for (i = 0; i < c.frames; i++) {
+		for (channel = 0; channel < sink->stream.channels; channel++) {
+			cd->buf_in[buff_frag] = *(int32_t *)audio_stream_read_frag_s32(&source->stream, buff_frag);
+
+			buff_frag++;
+		}
+	}
+	dcache_writeback_region(cd->buf_in, c.frames * sink->stream.channels * sizeof(int32_t));
```

Comment:
Can you use the buffer_acquire() API to get and put buffers. This does the cache management automatically for modules that share buffers across multiple cores (and avoids the WB when the buffers are on the same core).

Reply:
Not sure if this can apply here, these buffers are simple internal mem ptrs (not struct comp_buffer) because processing functions needs a continuous buffer as input/output,
but anyway I will check if this can work without writeback calls, it should

Reply:
ah, if there are internal only then its fine (as they wont be shared with other cores).